lib/ansible/plugins/filter/core.py

```diff
@@ -294,25 +294,30 @@ def mandatory(a):
 
 
 def combine(*terms, **kwargs):
-    recursive = kwargs.get('recursive', False)
-    if len(kwargs) > 1 or (len(kwargs) == 1 and 'recursive' not in kwargs):
-        raise AnsibleFilterError("'recursive' is the only valid keyword argument")
-
-    dicts = []
-    for t in terms:
-        if isinstance(t, MutableMapping):
-            recursive_check_defined(t)
-            dicts.append(t)
-        elif isinstance(t, list):
-            recursive_check_defined(t)
-            dicts.append(combine(*t, **kwargs))
-        else:
-            raise AnsibleFilterError("|combine expects dictionaries, got " + repr(t))
+    recursive = kwargs.pop('recursive', False)
+    list_merge = kwargs.pop('list_merge', 'keep')
+    if kwargs:
+        raise AnsibleFilterError("'recursive' and 'list_merge' are the only valid keyword argument")
 
-    if recursive:
-        return reduce(merge_hash, dicts)
-    else:
-        return dict(itertools.chain(*map(iteritems, dicts)))
+    # allow the user to do `[dict1, dict2, ...] | combine`
+    dictionaries = flatten(terms, levels=1)
+
+    # recursively check that every elements are defined (for jinja2)
+    recursive_check_defined(dictionaries)
+
+    if not dictionaries:
+        return {}
+
+    if len(dictionaries) == 1:
+        return dictionaries[0]
+
+    # merge all the dicts so that the dict at the end of the array have precedence
+    # over the dict at the beginning
+    result = dictionaries[0]
+    for dictionary in dictionaries[1:]:
```

Comment:
@bcoca I think it would be better to build the result from right to left as there is a huge probability that the leftmost dict will be the bigger one (as the leftmost will be the "default" and the other one will be the "patches"), and merge_hash create a copy of it's first argument.
But it add complexity. So I prefer to ask you before I made the change: should I build the result from left to right ? or right to left ?